src/vm/virtualMachine.ts

```diff
@@ -83,9 +91,83 @@ export class VirtualMachine {
     if (op1Error !== undefined) {
       return { value: undefined, error: op1Error };
     }
-    const op1 = this.segments.memory.get(op1Addr);
+    let op1 = this.segments.memory.get(op1Addr);
 
-    return { value: undefined, error: undefined };
+    // If op0 is undefined, then we can deduce it from the instruction, dst and op1
+    // We also deduce the result based on the result logic
+    if (op0 === undefined) {
+      const { value: deducedValues, error: deducedOp0Error } = this.deduceOp0(
+        instruction,
+        dst,
+        op1
+      );
+      if (deducedOp0Error !== undefined) {
+        return { value: undefined, error: deducedOp0Error };
+      }
+      const [deducedOp0, deducedRes] = deducedValues;
+      if (deducedOp0 !== undefined) {
+        this.segments.memory.insert(op0Addr, deducedOp0);
+      }
+      res = deducedRes;
+    }
+
+    // If op1 is undefined, then we can deduce it from the instruction, dst and op0
+    // We also deduce the result based on the result logic
+    if (op1 === undefined) {
+      const { value: deducedValues, error: deducedOp1Error } = this.deduceOp1(
+        instruction,
+        dst,
+        op0
+      );
+      if (deducedOp1Error !== undefined) {
+        return { value: undefined, error: deducedOp1Error };
+      }
+      const [deducedOp1, deducedRes] = deducedValues;
+      if (deducedOp1 !== undefined) {
+        this.segments.memory.insert(op1Addr, deducedOp1);
+      }
+      if (res === undefined) {
+        res = deducedRes;
+      }
+    }
+
+    // If res is still undefined, then we can compute it from op0 and op1
+    if (res === undefined) {
+      const { value: computedRes, error: computedResError } = this.computeRes(
+        instruction,
+        op0 as MaybeRelocatable,
+        op1 as MaybeRelocatable
+      );
+      if (computedResError !== undefined) {
+        return { value: undefined, error: computedResError };
+      }
+      res = computedRes;
+    }
+
+    // If dst is undefined, then we can deduce it from the instruction and res
+    if (dst === undefined) {
+      const { value: deducedDst, error: deducedDstError } = this.deduceDst(
+        instruction,
+        res
+      );
+      if (deducedDstError !== undefined) {
+        return { value: undefined, error: deducedDstError };
+      }
+      if (deducedDst !== undefined) {
```

Comment:
can deducedDst be undefined here? we checked for errors, it feels weird to check again

Reply:
but deduceDst can return undefined error and value, you have no error but it couldn't deduce the value